### PR TITLE
[CALCITE-3872] Simplify expressions with unary minus

### PR DESCRIPTION
**Pull request by liyafan82**
Support simplifying expression - ( - ( x ) ) as x.

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java [hunk 1]
```diff
@@ -2265,8 +2265,8 @@ private void assertTypeAndToString(
     checkSimplify(coalesce(gt(nullInt, nullInt), trueLiteral),
         "true");
     checkSimplify(coalesce(unaryPlus(nullInt), unaryPlus(vInt())),
-        "+(?0.int0)");
-    checkSimplifyUnchanged(coalesce(unaryPlus(vInt(1)), unaryPlus(vInt())));
+        "?0.int0");
```
Comment (ritesh-kapoor) [inline core/src/test/java/org/apache/calcite/rex/RexProgramTest.java]: Should we add more test cases which covers decimal as well?

Comment (liyafan82) [reply core/src/test/java/org/apache/calcite/rex/RexProgramTest.java]: Sounds good. I have added some test cases for decimal. Please take a look.

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java [hunk 2]
```diff
@@ -2265,12 +2265,35 @@ private void assertTypeAndToString(
...
     checkSimplify(coalesce(vInt(), nullInt, vInt(1)),
         "COALESCE(?0.int0, ?0.int1)");
+
+    checkSimplify(coalesce(vDecimalNotNull(), vDecimal()), // first arg not null
+        "?0.notNullDecimal0");
+    checkSimplifyUnchanged(coalesce(vDecimal(), vDecimalNotNull()));
+    checkSimplify(coalesce(vDecimal(), vDecimal()), // repeated arg
+        "?0.decimal0");
+    checkSimplify(coalesce(vDecimalNotNull(), vDecimalNotNull()), // repeated arg
+        "?0.notNullDecimal0");
```
Comment (chunweilei) [inline core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:2281]: Could you please put the comment on the top? Also squash commits.

**Comment by vlsi** [comment]
Does this change support -(-(min_integer)) and min_long cases?

I guess it should check the data type otherwise it will corrupt data.

File: core/src/test/java/org/apache/calcite/rex/RexProgramTest.java (hunk 2 above)
Comment (liyafan82) [reply core/src/test/java/org/apache/calcite/rex/RexProgramTest.java:2287]: Done. Thank you.

**Comment by liyafan82** [comment]
> Does this change support -(-(min_integer)) and min_long cases? I guess it should check the data type otherwise it will corrupt data.

@vlsi Thanks for the good question.
This should not be a problem, as we have already discussed in the jira. Please see https://issues.apache.org/jira/browse/CALCITE-3872

**Comment by DonnyZone** [comment]
@liyafan82 could you please rebase the pr?

**Comment by liyafan82** [comment]
> @liyafan82 could you please rebase the pr?

@DonnyZone Rebased. Thanks for your kind reminder.

**Pull request merged**